homeassistant/components/openplantbook/__init__.py

```diff
@@ -0,0 +1,135 @@
+            entry.data.get("client_id"), entry.data.get("secret")
+        )
+    if ATTR_SPECIES not in hass.data[DOMAIN]:
+        hass.data[DOMAIN][ATTR_SPECIES] = {}
+
+    async def get_plant(call):
+        species = call.data.get(ATTR_SPECIES)
+        if species:
+            # Here we try to ensure that we only run one API request for each species
+            # The first process creates an empty dict, and access the API
+            # Later requests for the same species either wait for the first one to complete
+            # or they returns immediately if we already have the data we need
+            _LOGGER.debug("get_plant %s", species)
+            if species not in hass.data[DOMAIN][ATTR_SPECIES]:
+                _LOGGER.debug("I am the first process to get %s", species)
+                hass.data[DOMAIN][ATTR_SPECIES][species] = {}
+            elif "pid" not in hass.data[DOMAIN][ATTR_SPECIES][species]:
+                # If more than one "get_plant" is triggered for the same species, we wait for up to
+                # 10 seconds for the first process to complete the API request.
+                # We don't want to return immediately, as we want the state object to be set by
+                # the running process before we return from this call
+                _LOGGER.debug(
+                    "Another process is currently trying to get the data for %s...",
+                    species,
+                )
+                wait = 0
+                while "pid" not in hass.data[DOMAIN][ATTR_SPECIES][species]:
+                    _LOGGER.debug("Waiting...")
+                    wait = wait + 1
+                    if wait == 10:
+                        _LOGGER.error("Giving up waiting for OpenPlantBook")
+                        return False
+                    await asyncio.sleep(1)
+                _LOGGER.debug("The other process completed successfully")
+                return True
+            elif datetime.now() < datetime.fromisoformat(
+                hass.data[DOMAIN][ATTR_SPECIES][species]["timestamp"]
+            ) + timedelta(hours=CACHE_TIME):
+                # We already have the data we need, so let's just return
+                _LOGGER.debug("We already have cached data for %s", species)
+                return True
+
+            plant_data = await hass.data[DOMAIN][ATTR_API].get_plantbook_data(species)
+            if plant_data:
+                _LOGGER.debug("Got data for %s", species)
+                plant_data["timestamp"] = datetime.now().isoformat()
+                hass.data[DOMAIN][ATTR_SPECIES][species] = plant_data
+                attrs = {}
+                for var, val in plant_data.items():
+                    attrs[var] = val
+                entity_id = async_generate_entity_id(
+                    f"{DOMAIN}" + ".{}", plant_data["pid"], current_ids={}
+                )
+                hass.states.async_set(entity_id, plant_data["display_pid"], attrs)
```

Comment:
I'm not familiar with anything in the plant area so take this comment with a grain of salt, but it feels unusual that services are creating and destroying entities. In the light/switch/sensor/etc world it'd be more traditional for the config_entry to know about what entities will be created and for them all to be created during setup.